hasktorch/src/Torch/Typed/Functional.hs

```diff
@@ -3905,8 +3905,24 @@ median' input = unsafePerformIO $ ATen.cast3 ATen.Managed.median_tlb
 -- argsort :: Tensor device dtype shape -> Int -> Bool -> Tensor device dtype shape
 -- argsort _input _dim _descending = unsafePerformIO $ (ATen.cast3 ATen.Managed.argsort_tlb) _input _dim _descending
 
--- topk :: Tensor device dtype shape -> Int -> Int -> Bool -> Bool -> (Tensor device dtype shape,Tensor device dtype shape)
--- topk _input _k _dim _largest _sorted = unsafePerformIO $ (ATen.cast5 ATen.Managed.topk_tllbb) _input _k _dim _largest _sorted
+
+-- | Returns the k largest (if largest is `True`) elements of the given input tensor along a given dimension.
+--
+-- >>> topk @3 @1 (ones :: CPUTensor 'D.Float '[2,3]) True True
+-- (Tensor Float [2,3] [[ 1.0000   ,  1.0000   ,  1.0000   ],
+--                     [ 1.0000   ,  1.0000   ,  1.0000   ]],Tensor Int64 [2,3] [[ 0,  1,  2],
+--                     [ 0,  1,  2]])
+topk 
+  :: forall k dim shape dtype device 
+   . (KnownNat k, KnownNat dim, All KnownNat shape, DimOutOfBoundCheck shape dim, FromJust (ExtractDim dim shape) >= k) 
+   => Tensor device dtype shape 
+   -> Bool -- ^ if we're returning the top k largest (or top k smallest)
+   -> Bool -- ^ if the resulting k elements are themselves sorted
+   -> (Tensor device dtype (FromJust (ReplaceDim dim shape k)), Tensor device dtype (FromJust (ReplaceDim dim shape 1)))
```

Comment:
nice!
alternatively, you could have defined a `TopKCheck` type family, maybe similar to `ChunkCheck`,
```Haskell
type family TopKCheck (shape :: [Nat]) (dim :: Nat) (result :: Maybe a) :: a where
  TopKCheck shape dim Nothing       = DimOutOfBound shape dim
  TopKCheck _     _   (Just result) = result
```
with
```Haskell
type TopK k shape dim = TopKCheck shape dim (ReplaceDim dim shape k)
```
This allows for customization of the error message the compiler produces, to the benefit of the user.

Reply:
Ah, yes, that's helpful, I wasn't sure what the threshold was for introducing new type-level ops.  I'll add that in `Functional.hs`